src/metabase/search/postgres/index.clj

```diff
@@ -0,0 +1,124 @@
+(ns metabase.search.postgres.index
+  (:require
+   [clojure.string :as str]
+   [honey.sql :as sql]
+   [honey.sql.helpers :as sql.helpers]
+   [toucan2.core :as t2]))
+
+(def ^:private active-table :search_index)
+
+(def ^:private pending-table :search_index_next)
+
+(def ^:private retired-table :search_index_retired)
+
+(defonce ^:private initialized? (atom false))
+
+(defonce ^:private reindexing? (atom false))
+
+(def ^:private tsv-language "english")
+
+(defn- exists? [table-name]
+  (t2/exists? :information_schema.tables :table_name (name table-name)))
+
+(defn- drop-table! [table-name]
+  (boolean
+   (when (exists? table-name)
+     (t2/query (sql.helpers/drop-table table-name)))))
+
+(defn- rename-table! [old new]
+  (when (and (exists? old) (not (exists? new)))
+    (-> (sql.helpers/alter-table old)
+        (sql.helpers/rename-table new)
+        t2/query)))
+
+(defn create-pending!
+  "Create a non-active search index table."
+  []
+  (when (not @reindexing?)
+    (when-not (exists? pending-table)
+      (-> (sql.helpers/create-table pending-table)
+          (sql.helpers/with-columns
+            [[:id :bigint [:primary-key] [:raw "GENERATED BY DEFAULT AS IDENTITY"]]
+             ;; entity
+             [:model_id :int :not-null]
+             [:model [:varchar 254] :not-null] ;; TODO find the right size
+             ;; search
+             [:search_vector :tsvector :not-null]
+             ;; scoring related
+             [:model_rank :int :not-null]
+             ;; permission related entities
+             [:collection_id :int]
+             [:database_id :int]
+             [:table_id :int]
+             ;; filter related
+             [:archived :boolean]])
+          t2/query)
+
+      (t2/query
```

Comment:
do you think it's worth creating an unique index on `(model, model_id)` to prevent duplications?

Reply:
I thought about this, but balked for now as I don't want to deal with the error handling. Will probably change to an upsert when it's out of the prototype phase.

Reply:
Also, kinda dumb reason, but for now I've been intentionally populating it multiple times to get a sense for the scaling of the index.